packages/@lwc/template-compiler/src/parser/parser.ts

```diff
@@ -134,33 +158,147 @@ export default class ParserCtx {
      * until it finds one where predicate returns true.
      */
-    findSibling<A extends ParentNode>(predicate: (node: ParentNode) => node is A): A | null {
-        const currentScope = this.currentScope() || [];
-        const sibling = currentScope.find(predicate);
-        return sibling || null;
+    findInCurrentElementScope<A extends ParentNode>(
+        predicate: (node: ParentNode) => node is A
+    ): A | null {
+        const currentScope = this.currentElementScope() || [];
+        return currentScope.find(predicate) || null;
     }
 
-    beginScope(): void {
-        this.scopes.push([]);
+    beginElementScope(): void {
+        this.elementScopes.push([]);
     }
 
-    endScope(): void {
-        this.scopes.pop();
+    endElementScope(): ParentNode | undefined {
+        const scope = this.elementScopes.pop();
+        return scope ? scope[0] : undefined;
     }
 
-    addNodeCurrentScope(node: ParentNode): void {
-        const currentScope = this.currentScope();
+    addNodeCurrentElementScope(node: ParentNode): void {
+        const currentScope = this.currentElementScope();
 
         /* istanbul ignore if */
         if (!currentScope) {
-            throw new Error("Can't invoke addNodeCurrentScope if there is no current scope");
+            throw new Error("Can't invoke addNodeCurrentElementScope if there is no current scope");
         }
 
         currentScope.push(node);
     }
 
-    private currentScope(): ParentNode[] | undefined {
-        return this.scopes[this.scopes.length - 1];
+    hasSeenSlot(name: string): boolean {
+        return this.hasSeenSlotInAncestorIfTree(name);
+    }
+
+    addSeenSlot(name: string): void {
+        const currentSeenSlots = this.seenSlotsFromAncestorIfTree();
+        if (currentSeenSlots) {
+            currentSeenSlots.add(name);
+        } else {
+            this.seenSlots.add(name);
+        }
+    }
+
+    private currentElementScope(): ParentNode[] | undefined {
+        return this.elementScopes[this.elementScopes.length - 1];
+    }
+
+    beginSiblingScope() {
+        this.siblingScopes.push({
+            ancestorIfContext: this.currentIfContext() || this.ancestorIfContext(),
+        });
+    }
+
+    endSiblingScope() {
+        this.siblingScopes.pop();
+    }
+
+    // Next goal is to try and move some of the "business logic" from here out into the parser's index.ts itself
+    // so that ParserCtx can be more of an utility than handling specific logic for how to handle the nodes.
+    beginIfChain(node: IfBlock) {
+        const currentSiblingContext = this.currentSiblingContext();
+        if (!currentSiblingContext) {
+            throw new Error('Cannot invoke beginIfChain if there is currently no sibling context');
+        }
+
+        // An if block always starts a new chain.
+        if (this.isParsingIfBlock()) {
+            this.endIfChain();
+        }
```

Comment:
From a pure separation of concerns perspective shouldn't this code be moved to the `parseIfBlock` in `index.ts`? The  `beginIfChain` method should on the other throw if the parser is already in the middle of parsing an `ifBlock`.

Reply:
I've moved it over and added a more descriptive error message.